### ISSUE TITLE
[Demand] Cryptocurrency Features (Stage 1)

### ISSUE THREAD

**Issue by neruthes** [DimensionDev/Maskbook#634]
## Metadata

Field | Value
--- | ---
Previous Stage | ...
Figma Link | [Dashboard "Wallets" Tab](https://www.figma.com/file/oFrOUZmDnOWL2pu6SNf4P7/Dashboard-v2?node-id=2122%3A1) / [Composing](https://www.figma.com/file/nDyLQp036eHgcgUXeFmNA1/Post-Composition-v1?node-id=0%3A1) / [Viewing](https://www.figma.com/file/pOQaNCE3cmeA3mxrmUKHX9/Twitter-Red-Packet-Demo?node-id=1%3A2)
Expected UX prototype due | 2020-01-10
Actual UX prototype release | 2020-01-10
Expected engineering due | 2020-01-17 (FRI)
Corresponding Pull Request | -
Completion Version | -

--------

## Catalogue

- Dashboard "Wallets" Tab
  - Wallet Card
  - AddWallet Section
  - Red Packets History
  - Red Packet Detail Page
- Red Packet Card
- Post Composing
  - List of Plugins
  - Plugin Modal
- Red Packet Plugin Modal Content
  - Create New Red Packet
  - Select Existing Red Packet
- Post Affixing Canvas (PAC)
- Red Packet PAC

--------

### Components

- Wallet Card
- AddToken Modal
- SendRedPacket Modal
- RedPacketsHistory Modal: List
- RedPacketsHistory Modal: Detail

### Component: Wallet Card

<img width="682" alt="Screen Shot 2020-01-19 at 07 48 35" src="https://user-images.githubusercontent.com/2139261/72677058-1ded1600-3a90-11ea-8c27-f41cbf8ba5d8.png">

<img width="291" alt="Screen Shot 2020-01-19 at 07 48 47" src="https://user-images.githubusercontent.com/2139261/72677061-234a6080-3a90-11ea-94a4-90b5991d2458.png">

#### Area: Name

Name of the corresponding Persona.

#### Area: Wallet Address

Generated by the public key according to related Ethereum standards. Can be copied into the clipboard.

#### Area: Red Packets History Button

Click to open RedPacketsHistory modal.

#### Area: Send Red Packet Button

Click to open SendRedPacket modal.

#### Area: List of Tokens

Each cell is a balance view of the token, indicating Token Symbol, Token Name, and currently available balance (calculated by real on-chain balance subtracting locally frozen balance).

The precision of the balance number remains to be specified. The current draft strategy: `if (balance < 10000) { display literal value with 5 significant digits } else { display a string like "123K" or "1.23M" or "12.3B", totaling 3 significant digits }`.

The tokens are sorted by Unicode dictionary order, ascending by codepoint.

When the Persona tries opening a Red Packet which is denominated in a token which is not in the list, the denominated token is automatically added into the list.

#### Area: Edit List Button

Click to change the balance values into "Delete" text label buttons (same to the ones used in `PersonaCard/ProfileCell/Connected`) and itself to a `Done` text label button, which upon clicking makes the list back to a for-display state. ETH itself may not be deleted.

#### Area: Add Token Button

Click to invoke AddToken modal.

### Component: AddWallet Section

<img width="657" alt="Screen Shot 2020-01-19 at 07 54 35" src="https://user-images.githubusercontent.com/2139261/72677132-f480ba00-3a90-11ea-8bb9-d6630e10379b.png">

### Component: AddToken Modal

<img width="709" alt="image" src="https://user-images.githubusercontent.com/2139261/72131085-2c8e4b80-3373-11ea-93ff-77b1e7150fd2.png">

The user may give a token contract to add it to the list of tokens.

### Component: SendRedPacket Modal

<img width="673" alt="image" src="https://user-images.githubusercontent.com/2139261/72580774-06b5f900-38d5-11ea-8cc8-6e1e9a308c0e.png">

The list contains all social networks which the current Persona has Profiles connected.

### Component: RedPacketsHistory Modal: List

<img width="315" alt="image" src="https://user-images.githubusercontent.com/2139261/72580813-26e5b800-38d5-11ea-9bdb-a3860479d25b.png">

The list contains all red packets which are created by the current Persona and every red packet which the current Persona tried to open (including "coming to late", "network error", and "cannot pay gas fee"). The list does not contain any red packet which the current Persona saw but did not try to open.

#### Area: Red Packet Card

See [DimensionDev/Tessercube-iOS#87-567398827](https://github.com/DimensionDev/Tessercube-iOS/issues/87#issuecomment-567398827).

### Component: RedPacketsHistory Modal: Detail

<img width="902" alt="image" src="https://user-images.githubusercontent.com/2139261/72592268-257ab680-38fa-11ea-9f8a-84a43cddad13.png">

#### Area: "Say Thanks" Button

The button will be visible when certain conditions are satisfied. Elaborations are defined at scenarios which invoke this modal.

#### Area: Source

This is the URL of the post where the current Persona initially tried opening.

Note that a Red Packet may appear in multiple Posts. All such Posts present the same Red Packet.

The scope of Red Packet is within Persona. Each Persona manage its relation with the same Red Packet. However, there are such exceptions:

- Number of collected shares: Whenever this field is updated, other Personae automatically receive the new number without initiating another query on the blockchain.

#### Area: Message

This is the attached message from the sender.

#### Area: List of Openers

Each cell is a summary of the opener.

The name is the nickname of the receiving Persona, which is specified when a collection operation is executed on the Ethereum network. However, if the receiving Persona is a local Persona (defined as whose private key exists in the local database), this field will always show the current nickname of this Persona, instead of the on-blockchain name.

The wallet address is the wallet address of the receiving Persona, which is by the public key according to related Ethereum standards.

The number on the right is the amount the receiver collected. The serialization strategy conform to the one used in `WalletCard/ListOfTokens`.

----------

## Part 2: Red Packet Card

[Figma Link](https://www.figma.com/file/yFI57ZoROuqtfWYNa2qgUc/Maskbook-UI-Kit?node-id=52%3A60)

#### Send

- Sending
- Network Error
- Failure, Cannot Afford Gas
- Base
- Open, Opening
- Open, Opened
- Open, Too Late
- Open, Network Error
- Refund Available
- Refunded

#### Receive

- Ready
- Network Error
- Failure, Cannot Afford Gas
- Open, Opening
- Open, Opened
- Open, Too Late
- Open, Network Error

----------

### List of Plugins

<img width="691" alt="image" src="https://user-images.githubusercontent.com/2139261/72134365-cb1eaa80-337b-11ea-91a0-e3411dca6985.png">

In the Composition Modal, there will be a list of Plugins. Currently, the only available plugin is Red Packet.

Each plugin is identified by a versioned package name (e.g. `com.maskbook.redpacket:1`, meaning version `1` of `com.maskbook.redpacket `). Multiple versions of the same package must not coexist.

In this list, if a plugin is preinstalled and starts with `com.maskbook.`, it should be displayed by its human-readable nickname instead of package name.

When the chip is clicked, a modal for the plugin will appear.

The modal must have an AppBar which has a "Close" button to exit the modal. The title of the modal is `Plugin: ${packageName.toLowerCase()}` (for ordinary plugins) or `Plugin: ${humanReadbleNickname}` (for `com.maskbook.*`)

The content of the modal determines how it operates inside.

Each plugin operates its isolated database tables and communicate with Maskbook and other plugins through to-be-defined protocols such as JSON RPC.

----------

### Tab: Create

<img width="665" alt="image" src="https://user-images.githubusercontent.com/2139261/72135072-809e2d80-337d-11ea-967b-6a56b4734f21.png">

Field | Description
--- | ---
Token | Dropdown selection, candidates are tokens in `WalletCard/ListOfTokens`.
Split Mode | Dropdown selection, candidates are `Average` and `Random`. This choice is lost when the modal is closed.
Amount | `Amount per Share` for Average mode. `Total Amount` for Random mode.
Shares | Must be positive integer.
Attached Message | See [DimensionDev/Tessercube-iOS#87](https://github.com/DimensionDev/Tessercube-iOS/issues/87) for its specifications.

The balance line will show the currently available balance (as defined afore) of the selected token, with supplemental ETH balance (same serialization strategy with modification, retaining only 2 significant digits instead).

The text of the "Create" button is `Pack ${TotalAmount.maxSignificantDigits(3)} ${TokenSymbol}`.

When the total amount of the Red Packet exceeds the balance, the "Create" button will be disabled, and the number and the token symbol in the balance line will be red and bold. A tooltip ("Insufficient balance") will appear above the "Create" button when the mouse cursor hovers it.

### Tab: Select

<img width="699" alt="image" src="https://user-images.githubusercontent.com/2139261/72135262-f3a7a400-337d-11ea-818e-395d4d72d6ff.png">

The user may click an option to decide which Red Packet is expected to be published in the post. The list shows the Red Packets which are successfully deployed are are valid and are created by the current persona and have at least 1 uncollected share (as far as Maskbook knows). Incoming Red Packets, expired Red Packets, and fully collected Red Packets will not appear in the list.

The first option in the list is to remove the existing Red Packet in the Post. If there is no Red Packet in the Post, this option will not appear.

----------

## Part 5: Post Affixing Canvas (PAC)

<img width="658" alt="image" src="https://user-images.githubusercontent.com/2139261/72136578-02438a80-3381-11ea-94e3-07774f2f9d2f.png">

When Maskbook detects plugin-specific data, the data will be passed to the post content interpreter of the plugin, rendering an HTML document which Maskbook will put in a canvas (not a `<canvas>` tag) beneath the decrypted post content. The width of the canvas is always 355 px (for screens which have odd pixels per row) or 356 px (for screens which have even pixels per row). When the screen is too narrow in terms of logical pixels, a downsampling will be made.

An operation by the user inside the canvas may affect the content of the canvas or open a modal, where the plugin is free to display any content on a full-page basis.

### Base

<img width="459" alt="image" src="https://user-images.githubusercontent.com/2139261/72418116-208afb00-3772-11ea-9dd5-db35f4ee15f2.png">

### Opening

When the user clicks the "Open Red Packet" button, the button will become disabled (text: `Opening...`), and Red Packet plugin background will try to collect a share of the Red Packet.

### Receive-Opened

If the Red Packet is one that the current Persona is already opened by the current Persona and is not created by the current persona, the "Open Red Packet" button will be replaced by a "Say Thanks" button. Also there should be a "View Details" borderless button, which upon clicking will open a CanvasSideRedPacketDetail modal, which looks perfectly similar to `RedPacketsHistoryModal/Detail-Alt` but the left button on the left of the AppBar will be replaced by a "Close" button and the "Source" field will be omitted.

"Say Thanks" button will appear only if the RedPacket is not created by the current Persona and is successfully opened by the current Persona.

<img width="483" alt="image" src="https://user-images.githubusercontent.com/2139261/72417535-03096180-3771-11ea-8121-7c35742c0437.png">

### Send-Opened

--------

## Notes

- None.

**Comment by neruthes** [comment]
This PRD is mostly ready for delivery. Reviews are requested.

cc @yisiliu @Jack-Works @SunriseFox @guanbinrui

**Comment by Jack-Works** [comment]
> Each plugin operates its isolated database tables

It is a good security feature but it's not easy in the current stage.

> communicate with Maskbook and other plugins through to-be-defined protocols such as JSON RPC.

The plugin system is a big project so we should make it work first then add the plugin system in future

------------------------------------------

Question: Can a single wallet address operate two type of crypto currency (like ETH & USDT)?

**Comment by Tedko** [comment]
For: Question: Can a single wallet address operate two type of crypto
currency (like ETH & USDT)?

Yes — as long as its on the same chain (here eth and USDT-on-eth-as-erc20)

Further more, you can derive them from the same private key (though it’s
not secure)

**Comment by neruthes** [comment]
> The plugin system is a big project so we should make it work first then add the plugin system in future

Well now you know why I started working on designing the new payload and the plugin system. ([I:b])

**Comment by yisiliu** [comment]
In the wallet card, how does it render the view if one has more than 2 tokens? They will appear on a list or directly append to the bottom?

**Comment by guanbinrui** [comment]
> (e.g. it has already been opened by the current Persona or is created by the current Persona)

It's strange the creator can not open the red packet which they created.

**Comment by Tedko** [comment]
> > (e.g. it has already been opened by the current Persona or is created by the current Persona)
> 
> It's strange the creator can not open the red packet which they created.

Good point

**Comment by neruthes** [comment]
Note:

Red Packet card UI may be subject to adjustments. Prioritize other components instead.

**Comment by neruthes** [comment]
> In the wallet card, how does it render the view if one has more than 2 tokens? They will appear on a list or directly append to the bottom?

I suppose that the length control of the list may be left to future stages. This feature does not look being worth the time working on, considering that our time is in extreme shortage.

**Comment by neruthes** [comment]
> > (e.g. it has already been opened by the current Persona or is created by the current Persona)
> 
> It's strange the creator can not open the red packet which they created.

Noted. I will review it soon.

**Comment by neruthes** [comment]
> > > (e.g. it has already been opened by the current Persona or is created by the current Persona)
> > 
> > 
> > It's strange the creator can not open the red packet which they created.
> 
> Noted. I will review it soon.

Just concluded that it was a mistake. Now corrected.

**Comment by neruthes** [comment]
Edit:

Now only 5 significant digits are expected in List of Tokens for balance amounts lower than 10000.

**Comment by neruthes** [comment]
**Amendment**:

----------

**Procedure of collecting a Red Packet**:

- See the Red Packet in the news feed or a post detail page.
- Click "Open Red Packet". The Red Packet card moves into "Receive-Opening" state, with the button becoming disabled with some kind of animation indicating that the system is responding.
- If `success`:
  - If `page has been closed already`:
    - Send the user a notification telling that the Red Packet has been collected, which upon clicking will direct the user to the detail page of the post where the Red Packet is located.
  - Else:
    - Let the user know that it has been opened, by automatically opening the corresponding RedPacketDetail modal in the page, where the user can easily know that it has been opened.
- Else:
  - If `page has been closed already`:
    - Notify the user that the collection was not successful, by sending a notification which upon clicking directs the user to `Dashboard / Wallets / RedPacketHistoryModal[ThisPersona] / RedPacketDetailView`, where the user can know that the collection was not successful.
  - Else:
    - Open RedPacketDetail modal in the page, where the user can know that the collection was not successful.

**Related Questions**:

- Is it feasible to get the PostLocator of a post without actually visiting the page? cc @Jack-Works

**Comment by neruthes** [comment]
**Discussion Summary:**

Cryptocurrency features on Facebook will be delayed till #397 is finished.

**Comment by neruthes** [comment]
**Edit:**

Fixed a minor detail in `RedPacketDetailModal / List of Openers`.

**Comment by SunriseFox** [comment]
> Fixed a minor detail in `RedPacketDetailModal / List of Openers`.

Better demonstrate it right here as I have no way to browse the proto's catalog nor diff the changes.

**Comment by neruthes** [comment]
> > Fixed a minor detail in `RedPacketDetailModal / List of Openers`.
> 
> Better demonstrate it right here as I have no way to browse the proto's catalog nor diff the changes.

Added.

**Comment by neruthes** [comment]
**Edit:**

-----------

Red Packet plugin modal now has multiple steps when creating a Red Packet.

<img width="1394" alt="image" src="https://user-images.githubusercontent.com/2139261/72595823-98882b00-3902-11ea-9517-735fa5a2936f.png">

During the creation procedure (rendering as S.a.1), the modal content tells Plugin Modal Manager that `requestConfirmationBeforeClose` is set to be `true`.

With `requestConfirmationBeforeClose` being `true`, when the user clicks "Close" button in the AppBar of the modal, the modal will prompt a confirmation dialogue.

```
# Close Modal?

As requested by this plugin, you need to confirm that you want to close this modal, before this modal is closed.

[Cancel]  [OK]
```

**Comment by neruthes** [comment]
Any information for latest progress?

**Comment by neruthes** [comment]
**Update:**

Since now, Wallets are Personae are absolutely decoupled. All previously written coupling logics are revoked. Updates will be pushed into the main article on a case-by-case basis.

**Comment by SunriseFox** [comment]
Note: time is short

Maybe no time to make adjustments

**Comment by neruthes** [comment]
**Update:**

Since Facebook support is delayed and Wallet is decoupled from Persona, this modal will have 1 and only 1 button, "Twitter.com".

<img width="743" alt="Screen Shot 2020-01-19 at 07 49 51" src="https://user-images.githubusercontent.com/2139261/72677073-49700080-3a90-11ea-90b0-1b8688c1af78.png">

**Comment by neruthes** [comment]
**Update:**

Added:

- AddWallet Section
- Create Wallet
- Backup Wallet
- Import Wallet

---------

## AddWallet Section

<img width="661" alt="image" src="https://user-images.githubusercontent.com/2139261/72677241-30684f00-3a92-11ea-9bf3-86afc6be24c5.png">

## Create Wallet

<img width="675" alt="image" src="https://user-images.githubusercontent.com/2139261/72677247-4544e280-3a92-11ea-8928-5ff553d241a2.png">

## Backup Wallet

<img width="703" alt="image" src="https://user-images.githubusercontent.com/2139261/72677259-57bf1c00-3a92-11ea-8337-472ec30a23ba.png">

## Import Wallet

<img width="657" alt="image" src="https://user-images.githubusercontent.com/2139261/72677267-69a0bf00-3a92-11ea-83a7-df116d963993.png">

**Comment by neruthes** [comment]
**Note:**

<img width="685" alt="image" src="https://user-images.githubusercontent.com/2139261/72726658-c5497600-3b80-11ea-9d81-de2c5c45fba2.png">

Only when current token is not ETH shall this piece of text (in parentheses) be shown.

**Comment by neruthes** [comment]
## Question

- How can the first wallet come into existence?

## Possible Solutions

- Require the user to create one when opening a Red Packet or creating a Red Packet.
- Automatically generate one upon installation.

**Comment by neruthes** [comment]
## Note

When composing a post, if there is a Red Packet attached, only "Everyone" is allowed to be the recipients. The user may not manually change recipients.

This restriction will be removed in future.

**Comment by neruthes** [comment]
Looks like many of this issue have been implemented and released. Why does this issue remain open? Should any pull request have closed this issue?

**Comment by neruthes** [comment]
## Discussion Summary

Should have been closed by #672.

**Comment by guanbinrui** [comment]
Close by #672
